### DISCUSSION TITLE
<Link/> sometime stop working with app router in production

### DISCUSSION THREAD

**Discussion by hiiamwin** [vercel/next.js#57565] [App Router] [answered]
I am till learning nextjs.
I was hoping to get a bit of insight into a bug I'm experiencing with my nextjs application. Every once and a while all the links on my page seem to stop working.

If I hover over the links I can see the expected page in the bottom left of my browser but clicking doesn't navigate to the page and the url doesn't change too. After a refresh or wait a short period of time , the links usually start working again but it definitely a huge hit to usability for the average user.

Has anyone experienced a similar issue/ have any insight as to what might be causing this to occur?

Any help is greatly appreciated :)

**Comment by vignesh-gupta** [comment]
Hey @hiiamwin, Could please attach some video or a reproducing steps with the piece of code that would be helpful. Apart from that, if you are using `next/link` then potential reason could be that your navigated page is having a expensive (time taking) process like DB or API fetching. 

To identify this- 
1. Try adding a [loading.tsx](https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming) page along side your `page.tsx`
2. If not the page then try covering the component with [Suspence Component](https://react.dev/reference/react/Suspense) to load it faster.


Here is a good news for you - Yesterday [Next 14](https://nextjs.org/blog/next-14) has been launched with a game changing feature [Partial Prerendering (Preview)](https://nextjs.org/blog/next-14#partial-prerendering-preview) that should help in this.



**Hope it help. Happy Coding!**

**Reply by hiiamwin** [reply]
thanks sir

**Reply by vignesh-gupta** [reply]
Hey @hiiamwin if the answer works for you, Please mark as answered so that if someone else tries to find similar they can refer to this.

**Reply by ako-v** [reply]
I am using Next v14.2.5, I had the same issue, that once in a while all the links would stop working! using prefetch false did not work, but this solution worked for me. Thanks.

**Comment by aziyatali** [comment]
It's due to the caching. You can check the [this](https://nextjs.org/docs/app/building-your-application/caching#opting-out-2) for details.

Or you can watch this [video](https://www.youtube.com/watch?embeds_widget_referrer=https%3A%2F%2Fmedium.com%2F%40CraftedX%2Fwhy-i-hate-next-js-13-caching-4cd5a31d3a78&embeds_referring_euri=https%3A%2F%2Fcdn.embedly.com%2F&embeds_referring_origin=https%3A%2F%2Fcdn.embedly.com&source_ve_path=MTM5MTE3LDI4NjY0LDE2NDUwMw&feature=emb_share&v=_yhSh4g0NSk&ab_channel=Joshtriedcoding) to understand it better

There are several work-arounds, so you can try them by searching for your needs.

What we have done was trying not to use Link as it is caching the request.
One way could be switching to a tag of html.
Others you can use router but it still sucks sometimes. So you better refresh it(router.refresh()) before pushing as refresh seem to refetch the data again by disabling caching.

**Comment by hiiamwin** [comment, marked as answer]
i try to set prefetch to false and it work

**Reply by prasad-pul** [reply]
This really helped (after almost 15 hours of head scratching / debugging!!).

I wish Nextjs team provide more clarity / documentation around this. Why did this work??

For anyone interested here is the workflow / navigation steps that I had which broke before setting the `prefetch = {false}`

1. Navigate to the home page
2. There a bunch of sub-links (through the Nextjs `Link` component)
3. Clicking on one of the link takes me to the `unauthenticated` version of that (sub) route
4. After I authenticate the above (`authenticated version`) route is rendered
5. Now if I navigate to a different link on the home page
6. And click the **link from Step 3** , `**NOTHING**` happens for **30 seconds**
7. After **30 seconds**, clicking the link from Step 3 works fine

The issue **noticed Step 6** was resolved by setting the `prefetch = {false}` on the links rendered on homepage. I don't know why??

**Reply by C-W-D-Harshit** [reply]
i have a very complex sidebar and the components change on particular routes, soo you know how complex it it and routing through link tag worked fine in dev mode {using Next: 14.1.3 app} but it dosent work in prod, idk why but setting the prefetch={false} just made it work somehow

**Reply by hilkiano** [reply]
This setting worked for me. `prefetch={false}` really fix my issue in production mode. FYI I am using `Link` component from `next-intl` v3.

**Comment by adrianszablowski** [comment]
In my case nothing helped, after clicking `<Link href='/' prefetch={false} />` the same thing happens.... The URL does not change and I have a white page (photo below). The URL stays as it was on the previous page.

![image](https://github.com/vercel/next.js/assets/111178975/1622bdab-a85a-4357-a424-102f8d9d6eea)

**Reply by JClackett** [reply]
Also running into this, except when I click the link, sometimes it just hangs, and now I get a load of rage clicks from users - super strange

**Comment by bugproof** [comment]
I have the same issue...

Here's my scenario:

```tsx 
<Link href="/dashboard">Dashboard</Link>
```
 inside server component.

Dashboard calls server action, that server action calls `fetch` to fetch a list from my backend.

Sometimes after page refresh it randomly stops working, hangs... I'm investigating this issue... Perhaps server actions are bugged for initial data fetches?


```tsx

export default async function DashboardPage() {
    const myPages = await getMyPages(); // server action which calls my  API using fetch

    return <>....</>
    ```

**Reply by VichetN** [reply]
I have the same senario to @bugproof 

Anyone resolve the problem?
I still face this issue in v14.1.10

added **prefetch={false}** doesn't solve anything

I am using with Authjs v5

**Reply by mp3por** [reply]
Same here

**Reply by Enkratia** [reply]
Be carefull with `router.refresh()`, `cookies()`, `headers()` and other things like this. It can force to redownload current page instead of visiting another page via `<Link />` or `router.push`.

**Comment by MazenSenih** [comment]
I just had the same issue, I'm using Next.js version 14.1.3, all my Link components stopped navigating suddenly, after investigating the console, I found a useEffect dependency that caused infinite loop, fixing that fixed the Link issues, so I suggest you to look closely what other code you recently changed, or any console messages and fixing them.

**Comment by HakanSungur** [comment]
I have the same problem. It's a terrible situation for Next that this problem hasn't been solved yet. I have the same problem in the dev environment.

**Reply by abhishekyadav-r360** [reply]
for me only in production even i use router.push() its also not work

**Comment by Barre** [comment]
I encountered the same navigation issue, which proved to be quite challenging to debug.

In my case, the problem occurred specifically when:

1. Navigating from the problematic page to another page.
2. Using the browser's back button to return.

This sequence would cause the router to completely break, but only in Firefox and only with a production build - it worked as expected in WebKit/Blink-based browsers.

The root cause ended up being that I used an `a` element instead of a `Link` in one place.

**Comment by VichetN** [comment]
@hiiamwin Can you re-open this issue? It still not  working in Next v14 and v15

**Comment by dreamy32** [comment]
Issue is back in newer versions. Fix this ASAP

Edit: Setting `prefetch={false}` doesn't fix the problem

**Comment by dongalor** [comment]
Same here. I've tried creating my own link component, that uses useRouter (from next/navigation) and it looks like it's working fine. Also I've created a demo with next 15 (in next 14.2.8 it doesn't) and it seems that it works out of the box (with Link component it fetches everything in server action good), but according to some comments here it's not 🤷‍♂️

const router = useRouter();
onClick -> router.push(href)

P.S. ChatGPT says:

_This behavior occurs because the Link component from next/link and useRouter's push method interact with the client-side routing system, which doesn't trigger a full page reload. Consequently, some cookies (like HttpOnly or server-only cookies) that are only accessible on the server won't be re-processed on the client navigation._

**Reply by dongalor** [reply]
Using prefetch={false} seems to work well. I am testing it now, and it looks good. In the Next.js 15 demo, I had no problems. Unfortunately, I can't update my project to version 15 yet. I wish version 14 had a similar improvement for convenience.

**Reply by dreamy32** [reply]
What version of Next.js 15 are you on? I am on version 15.0.3 and `prefetch={false}` didn't change anything. 

_ps: I am redirecting to a legacy React app, without nextjs support_

**Reply by dongalor** [reply]
"next": "15.0.3"

Also I don't really use prefetch there. 

Redirecting to external app might be another case.

Here's my demo I've made for my purposes.

https://github.com/dongalor/next-15-redirect-server-action

**Comment by jj2020driver** [comment]
In our project we have such utility as Osano, security management tool which block malicious third-party scripts. When initialized, it prevents NextJs v15 with Pages router to function properly with symptoms as described in this thread. Hope this helps someone.

**Comment by dreamy32** [comment]
Issue has been resolved for my project with v15.1.0

**Reply by jijiseong** [reply]
Does that mean it was resolved in version 15.1.0?

**Reply by kris-terziev** [reply]
At least for me upgrading to 15.1.0 did not fix the issue

**Reply by dreamy32** [reply]
I believe it was a misunderstanding on my part about how Link component works. After rereading the doc, it worked fine as expected.

**Comment by svaraborut** [comment]
Still present in `^15.1.4`.

During development, I have never noticed it, but in production, It seems stable at first but quickly navigating between links will reach a point where a load is skipped and then that link will remain broken for some time.

_I'm unfamiliar with Next core, but if I would have to guess: It all seems like a prefetch loading race condition that crashes the `<Link>` to get then resolved by the next rendering cycle._

We have disabled `prefetch` for now.

**Reply by Barre** [reply]
I observed the same, but only on Firefox.

**Comment by alexanderbnelson** [comment]
Prefetch or no, my link components are regularly breaking in the same way. It's brutal actually. It happens in dev and prod for me. What a mess. running next@15.1.6

**Comment by KishoreVB70** [comment]
I have the issue where both Link component and `router.push` are not working

**Comment by bill-zhanxg** [comment]
Same issue on next@15.2.0-canary.33 with reactCompiler and ppr enabled. I think this discussion should be upgraded to Issues.
It seems like when the user navigates via nav bar links and then click a link from the children that changes ONLY the searchParams will disable all links (when click on does nothing other than my progress bar moving: next13-progressbar). Setting all the Links' prefetch (including navBar links) to false does solve the problem but I do not want to set all prefetch to false and it should be at least documented if not a bug.

Also if the user refresh the page without clicking any other link before the searchParam changing Link it will work, and then all Links will work even when navigated from the Nav Bar links. It's a pretty weird bug that will prob need some effort to find the problem. It might be a logic check error or JS event loop problem (unlikely)

**Reply by twillhorn** [reply]
turning on experimental.clientSegmentCache fixed the ppr issue for me

**Comment by gerismumo** [comment]
having the same issue with next js 14.2.4

**Comment by joelpierre** [comment]
Its crazy that this is an issue and claims to have been "answered" yet we are all experiencing the problem still in various ways. `prefetch=false` doesn't work all the time either but it is noticeably better on occasion.

Refreshing the page solves it, until it hits again. I can see in the network tab that when the link is clicked the page is fetched and data returned but the view just doesn't change.

**Reply by gerismumo** [reply]
experiencing the same yet on my other next js applications it's working.

**Reply by dreamy32** [reply]
Have you tried deleting the .next folder?

**Reply by joelpierre** [reply]
Its not a local dev issue. Its the production build. My dev env is fine.

**Reply by halilhanbadem** [reply]
yes this problem is definitely ongoing and I have no idea how to fix it

**Reply by joelpierre** [reply]
Solution I've found works. And it's a terrible one in my opinion. Is to set a key on a client component as close to your layout.tsx as possible. I use the pathname specifically but beware what this means from a rerender POV but I think the issue is nothing client side changes in order to trigger a rerender 🤔

**Reply by halilhanbadem** [reply]
I observed that this issue occurs in places where `router.refresh()` is used. Therefore, instead of using `router.refresh()`, I replaced it with the native `window.location.reload();` call, and the issue was resolved.

What the user reported was that, on the login screen, even though the credentials were correct, the screen would remain unchanged, and this would happen occasionally. I investigated the issue and saw that the data was being loaded in the background, but the page wasn't refreshing. I then realized that I was using `router.refresh()` in my `auth` library. Although `router.refresh()` is intended to trigger an update after a server-side change, I believe there's a problem in this context.

As a result, the current implementation is working quite well. When I did some research on `router.refresh()`, I came across the following answer, and I think I made the right choice (though it's debatable): https://stackoverflow.com/a/77663070. It seems like the issue with `router.refresh()` is either related to the way I'm using it or it's a more general problem. Instead of diving deeper into the specifics of the problem, calling a native refresh is a simpler and effective workaround for now.

**Comment by ogheo** [comment]
I had same problem in nextjs 15 so for now disabled optimisticClientCache and set staleTimes to 0 like this:
```
  experimental: {
    optimisticClientCache: false,
    staleTimes: {
      dynamic: 0,
      static: 0,
    },
  },
```

**Reply by thenoobchad** [reply]
What file was this placed in?

**Reply by joelpierre** [reply]
nextjs config file I believe

**Comment by Stan-Stani** [comment]
Seeing this same problem. Is there a corresponding item in the issues? I'm on "next@npm:14.2.25 btw.

**Reply by Stan-Stani** [reply]
Our issue was that we set `queryParams` once our component initializes on a page, so if a user clicked a `<Link />` after the `router.replace()` was called but before it actually updated the url, the users navigation would get effectively cancelled by the replace. This is because Next.Js's app router methods have asynchronous effects even though the methods themselves aren't marked async.

**Comment by bolechen** [comment]
Hello everyone,

I'd like to share our debugging experience regarding a persistent client-side navigation 
issue, which might be related to the problems discussed here.

**Problem:**

In our Next.js 15.x application (specifically observed on `15.4.0-canary.110`), client-side
navigation would break after performing a search on the `/all` page. Specifically, after
navigating from the homepage to `/all` (via `<Link>`) and then executing a search (which
triggers `router.replace()`), all subsequent client-side navigation links would become
unresponsive. This issue was reproducible only on desktop in production builds.

**Initial Misleading Clues (and why they were ruled out):**

We extensively debugged various aspects, including:
*   Navbar component architecture (SSR/CSR split).
*   SWR data fetching configurations (`keepPreviousData`, `revalidateOnMount`).
*   CSS animations.
*   Specific client-side interactive components (e.g., DropdownMenu, ThemeToggle, Drawer).
*   All these were ruled out after meticulous, step-by-step elimination.

**Actual Root Cause:**

The root cause was identified as a conflict with **Next.js's Partial Prerendering (PPR) 
experimental feature**. When PPR was enabled, navigating client-side to a dynamic route 
(like `/all`, which uses `searchParams` and was later marked `export const dynamic = 
'force-dynamic'`) would seemingly corrupt the App Router's internal state. Subsequent 
`router.replace()` calls would then fail silently.

**Solution:**

1.  **Disabled PPR:** Set `ppr: false` in `next.config.ts`.
2.  **Marked Dynamic Route:** Added `export const dynamic = 'force-dynamic';` to `src/app/(
main)/all/page.tsx` to correctly declare it as a dynamic route (this also resolved a
`DYNAMIC_SERVER_USAGE` error).
3.  (Optional but recommended) Downgraded Next.js from a Canary version (`15.4.0-canary.110
`) to the latest stable release (`15.3.5`) to ensure overall stability.

After disabling PPR, the navigation issue was completely resolved. This suggests a deeper
interaction bug within the PPR implementation when combined with client-side navigation to
dynamic routes.

Hope this detailed experience helps others!

**Reply by bill-zhanxg** [reply]
This is very helpful. Thanks for the time spent debugging

**Reply by twillhorn** [reply]
Did you try turning on experimental.clientSegmentCache instead of disabling ppr? That worked for me.

**Comment by issam-seghir** [comment]
I'm on v15.5.4 and the issues still occur, the only solution that works for me is : `prefetch=false` on each Link component

**Comment by cbratschi** [comment]
Had the same issue on 15.5.5 and 15.5.4. After clicking a link and navigating back, the link no longer works. This is reproducible with all links on this page. After disabling prefetching it works normally. Some links had a small delay which is now gone too.

There were never any issues in dev mode. It just appeared in production builds.

Any way to debug this? Prefetching is a very useful feature but it has to work.

**Comment by WolfspiritM** [comment]
I also encounter this issue. When clicking links fast it suddenly stops working. The URL doesn't change anymore but calls are happening. For me it happens when switching pages fast AND having a redirect like this in the page the link is forwarding to:

```ts
import { redirect } from "next/navigation";
export default async function Page({ params }: { params: Promise<{ testId: string }> }) {
    const { testId } = await params;
    redirect(`${testId}/dashboard`);
    return null;
}
```

e.g. /test -> /test/dashboard

Then having two or more links going to /test, /test2, ...

I was able to dig deeper into it and the problem for me seems to be that the Router Component (app-router.tsx) fails to render (which also takes care via another component to update the URL). According to devtools it seems to break at the first "useMemo". Setting devtools to break on caught exceptions I get an error "Rendered more hooks than during the previous render." coming from the useMemo call. Not sure why this doesn't surface anywhere.

I wonder if it's related to:
https://github.com/facebook/react/issues/33580

And the use of "use(stateWithDebugInfo)" in "useActionQueue" if the page is not prefetched.

But it's strange that it works when not having the redirect inbetween but as the problem seems to be a timing issue it might be due to the "double" request.

**Comment by lucasdu4rte** [comment]
I'm having the same problem in an app with pagination and search params. Next.js 15.5.4. I already tried updating Next.js to 15.5.5, but the problem persists.

**Comment by V-iktor** [comment]
Same issue here. `<Links />` simply stop working.

I have a category drawer, every link there is generated on build time using `generateStaticParams`.

Clicking those links sometimes happen instantly, but other times takes 10-30 seconds to change the route.

Nothing happens, not even the URL in the browser changes when the link is clicked.

**Reply by dorji-dev** [reply]
> Same issue here. `<Links />` simply stop working.
> 
> I have a category drawer, every link there is generated on build time using `generateStaticParams`.
> 
> Clicking those links sometimes happen instantly, but other times takes 10-30 seconds to change the route.
> 
> Nothing happens, not even the URL in the browser changes when the link is clicked.

I am also having the same issue as yours. At times work instantly and at times takes forever. Also what I have noticed is that it seems to happen on tabs that's left open and if some minutes pass by without any activity, coming to the tab and clicking on any link takes forever to navigate. I am self hosting it by the way on digital ocean. Next = 16.1.0

**Comment by illepic** [comment]
In case this is helpful for anyone:

Edit: my initial comment about this being a serialization issue around Date objects is WRONG.

I was using the ShadCN Tab component. This component should ONLY exist in a "use client" component, but I had it in the middle of an SSR page. By re-architecting it to be "use client", there is no more Link navigation issues to this page.

**Comment by MatteoAlberghini** [comment]
Disabling prefetch makes this component useless, does anyone have a reason why Link just stops working sometimes? next16 still has the same issue (I'm running a fully static website with 2000 or so pages)

**Comment by trickyc0d3r** [comment]
Same fckn problem , next 16. My boss tought that im did it. Fck next!

**Comment by degibons** [comment]
Seeing the same thing here.

In my case, after the tab stays unused for some time, client-side navigation can just stop working. Links are still clickable visually, but clicking does nothing until a full page reload.

I’m on the latest Next.js (16.1.6) with App Router. prefetch is already disabled in my link wrapper (prefetch={false}), so that workaround does not help here.

At first I noticed it mostly on iPad, but I’ve also reproduced it on Windows 10 in Chrome, so it doesn’t seem to be Safari-only.

**Comment by ahmed-fawzy99** [comment]
Please re-open the issue.

Still happening on v16.1.6. Recreatable via opening a tab - leave it unused for some time (maybe hours) - come back - all navigation is not working.

The page is functional, I can see the links, just nothing happens upon clicking. No errors on console. On Network tab, I see a successful rsc fetch call, just no navigation happening.
I used a mixture of prefetched and non-prefetched links. Happens on all, regardless.

**Comment by minh97tt** [comment]
Still happening here.

Downgrade from `15.5.x` to `"next": "15.3.5"` would help.
Disable `prefetch` or update `next.config` doesn't work in my case.

**Comment by cbratschi** [comment]
A form action was prefetched in Next.js 16.2.6 and on submit was getting the 404 error page. Had to disable prefetch for this form too.
